plugins/amqp_compressed_proof_plugin/amqp_compressed_proof_plugin.cpp

```diff
@@ -0,0 +1,391 @@
+      }
+      else {
+         new_node.cmd_stream.splice(new_node.cmd_stream.end(), left.cmd_stream);
+         new_node.cmd_stream.splice(new_node.cmd_stream.end(), right.cmd_stream);
+         new_node.cmd_stream.emplace_back(combine_nodes());
+      }
+
+      return new_node;
+   }
+
+   std::vector<merkle_creator_node>   curent_level;
+};
+
+struct compressed_proof_generator_impl {
+   compressed_proof_generator_impl(chain::controller& controller, std::vector<compressed_proof_generator::result_callback_funcs>&& callbacks) :
+     callbacks(callbacks), irreversible(controller.get_read_mode() == chain::db_read_mode::IRREVERSIBLE) {
+      controller_connections.emplace_back(controller.irreversible_block.connect([this,&controller](const chain::block_state_ptr& bsp) {
+         on_irreversible_block(bsp, controller);
+      }));
+      controller_connections.emplace_back(controller.applied_transaction.connect([this](std::tuple<const chain::transaction_trace_ptr&, const chain::packed_transaction_ptr&> t) {
+         on_applied_transaction(std::get<0>(t));
+      }));
+      controller_connections.emplace_back(controller.accepted_block.connect([this,&controller](const chain::block_state_ptr& p) {
+         on_accepted_block(p, controller);
+      }));
+      controller_connections.emplace_back(controller.block_start.connect([this](uint32_t block_num) {
+         on_block_start(block_num);
+      }));
+   }
+
+   void on_applied_transaction(const chain::transaction_trace_ptr& trace) {
+      if(!trace->receipt)
+         return;
+      //only executed & delayed transaction traces would make it in to action_mroot
+      if(trace->receipt->status != chain::transaction_receipt::status_enum::executed &&
+         trace->receipt->status != chain::transaction_receipt::status_enum::delayed)
+         return;
+
+      if(chain::is_onblock(*trace))
+         onblock_trace.emplace(trace);
+      else
+         cached_traces[trace->id] = trace;
+   }
+
+   void on_accepted_block(const chain::block_state_ptr& bsp, const chain::controller& controller) {
+      std::set<action_entry> action_entries_this_block;
+
+      if(onblock_trace) {
+         for(const chain::action_trace& at : (*onblock_trace)->action_traces)
+            action_entries_this_block.emplace(at);
+      }
+      for(const chain::transaction_receipt& r : bsp->block->transactions) {
+         transaction_id_type id;
+         if(r.trx.contains<chain::transaction_id_type>())
+            id = r.trx.get<chain::transaction_id_type>();
+         else
+            id = r.trx.get<chain::packed_transaction>().id();
+         auto it = cached_traces.find(id);
+         EOS_ASSERT(it != cached_traces.end(), chain::misc_exception, "missing trace for transaction ${id}", ("id", id));
+         for(const chain::action_trace& at : it->second->action_traces)
+            action_entries_this_block.emplace(at);
+      }
+      clear_caches();
+
+      if(irreversible)
```

Comment:
If the `db_read_mode` is `IRREVERSIBLE`, will there ever be a situation where `on_accepted_block` is not immediately followed by `on_irreversible_block`?

Reply:
I wasn't able to quite convince myself that removing all that is safe. I couldn't find a good way of testing all the cases in unit tests because, for example, I couldn't find a good way to switch between speculative & irreversible mode while the proof generator was attached to controller's signals (because all of tester's ctors init the controller).

Reply:
After playing with it some more, I can't find any problems just driving it via `on_accepted_block`